source/common/common/assert.cc

```diff
@@ -28,15 +32,87 @@ class ActionRegistrationImpl : public ActionRegistration {
   static std::function<void()> debug_assertion_failure_record_action_;
 };
 
+// This class implements the logic for triggering ENVOY_BUG logs and actions. Logging and actions
+// will be triggered with exponential back-off per file and line bug.
+class EnvoyBugRegistrationImpl : public ActionRegistration {
+public:
+  EnvoyBugRegistrationImpl(std::function<void()> action) {
+    ASSERT(envoy_bug_failure_record_action_ == nullptr,
+           "An ENVOY_BUG action was already set. Currently only a single action is supported.");
+    envoy_bug_failure_record_action_ = action;
+    counters_.clear();
+  }
+
+  ~EnvoyBugRegistrationImpl() override {
+    ASSERT(envoy_bug_failure_record_action_ != nullptr);
+    envoy_bug_failure_record_action_ = nullptr;
+  }
+
+  // This method is invoked when an ENVOY_BUG condition fails. It increments a per file and line
+  // counter for every ENVOY_BUG hit in a mutex guarded map.
+  // The implementation may also be a inline static counter per-file and line. There is no benchmark
+  // to show that the performance of this mutex is any worse than atomic counters. Acquiring and
+  // releasing a mutex is cheaper than a cache miss, but the mutex here is contended for every
+  // ENVOY_BUG failure rather than per individual bug. Currently, this choice reduces code size and
+  // has the advantage that behavior is easier to understand and debug, and test behavior is
+  // predictable.
+  // TODO(asraa): Re-evaluate implementation as usage increases or mutex contention is observed.
```

Comment:
I don't think this TODO is needed. It doesn't add any more information than the above comment.